[master < T1023-DX] Add meta_util module docs (#646)
memgraph/docs: mage/query-modules/python/meta-util.md

---
title: meta_util
sidebar_label: meta_util
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
import RunOnSubgraph from '../../templates/_run_on_subgraph.mdx';

export const Highlight = ({children, color}) => (
  <span
    style={{
      backgroundColor: color,
      borderRadius: '2px',
      color: '#fff',
      padding: '0.2rem',
    }}>
    {children}
  </span>

> );

Reviewer: E za sto je ovo molim te, bas sam se pital anedavno cemu to sluzi?

Author: A custom component that gives you background color (highlights text) in a table explaining the algorithm. In this case, the color is #FB6E00 (orange), but the component can be used in other places in different colors.
<img width="342" alt="Screenshot 2023-01-30 at 13 31 40" src="https://user-images.githubusercontent.com/61758502/215477759-d9996d1e-852b-44c8-9b5c-e5dbbd14b65e.png">

<!-- TODO -- update link -->
[![docs-source](https://img.shields.io/badge/source-meta_util-FB6E00?logo=github&style=for-the-badge)](https://github.com/memgraph/mage/blob/main/python/meta_util.py)

## Abstract

A module that contains procedures describing graphs on a meta-level.

| Trait               | Value                                                 |
| ------------------- | ----------------------------------------------------- |
| **Module type**     | <Highlight color="#FB6E00">**util**</Highlight>  |
| **Implementation**  | <Highlight color="#FB6E00">**Python**</Highlight>     |
| **Parallelism**     | <Highlight color="#FB6E00">**sequential**</Highlight> |

## Procedures

<RunOnSubgraph/>

### `schema(include_properties)`

> Returns a list of distict relationships connecting distinct nodes, that is, a graph schema. If `include_properties` is set to `true`, the graph schema will contain properties count information. 

Reviewer: ```suggestion
Returns a list of distinct relationships connecting distinct nodes, that is, a graph schema. If `include_properties` is set to `true`, the graph schema will contain additional information about properties. 
```

#### Input:

> * `include_properties: bool (default=false)` ➡ If set to `true`, the graph schema will include properties count information. 

Reviewer: See above
properties count doesn't mean much if you don't know what it's really about?

Reviewer: I see know you use it quite a lot throughout the documentation... but I'm still wondering if people will know what it is, and what you mean?

Author: I added a sentence at the beginning of the procedure description. Let me know if that clears things up a bit or should I go fix it all?

#### Output:

> * `nodes: List[Map]` ➡ list of distinct node objects with their count. If `include_properties` was set to `true`, the node object contains properties count too.
> * `relationships: List[Map]` ➡ list of distinct relationship objects with their count. If `include_properties` was set to `true`, the relationship object contains properties count too.

Reviewer: ```suggestion
* `nodes: List[Map]` ➡ List of distinct node objects with their count. If `include_properties` is set to `true`, the node object contains properties count too.
* `relationships: List[Map]` ➡ List of distinct relationship objects with their count. If `include_properties` is set to `true`, the relationship object contains properties count too.
```

#### Usage:
Get graph schema without properties count:
```cypher
CALL meta_util.schema() 
YIELD nodes, relationships 
RETURN nodes, relationships;
```

Get graph schema with properties count:
```cypher
CALL meta_util.schema(true) 
YIELD nodes, relationships 
RETURN nodes, relationships;
```

:::info

> It will show a graph view only in Memgraph Lab version >= 2.4.0. For earlier versions of the Memgraph Lab, `UNWIND` needs to be called on returned object properties nodes and edges.

Reviewer: ```suggestion
The queries above will return results in the graph view only in Memgraph Lab version >= 2.4.0. For earlier versions of the Memgraph Lab, call `UNWIND` on returned object properties nodes and edges.
```

  ]
}>
  <TabItem value="graph1">

  Create a graph by running the following Cypher query:

```cypher
CREATE (n:Person {name: "Kate", age: 27})-[:IS_FRIENDS_WITH]->(m:Person:Student {name: "James", age: 30, year: "second"})-[:STUDIES_AT]->(:University {name: "University of Vienna"})
WITH n, m
CREATE (n)-[:LIVES_IN]->(:City {name: "Zagreb"})<-[:LIVES_IN]-(m);
```
  </TabItem>

  <TabItem value="run1">

Once the graph is created, run the following code to call the `schema` procedure:

```cypher
CALL meta_util.schema() 
YIELD nodes, relationships 
RETURN nodes, relationships;
```

  </TabItem>

  <TabItem value="graph-result1">

The graph result of the `schema` procedure can be seen in Memgraph Lab, and it looks like this:
<!-- 
<div style={{textAlign:"center"}}> -->
<div className={"imgRow"}>
<div className={"imgColumn"}>
<img src={require('../../data/query-modules/python/meta_util/meta_result.png').default}/>
</div>
<div className={"imgColumn"}>
<img src={require('../../data/query-modules/python/meta_util/meta_result_count_without_props.png').default}/>
</div>
</div>
  </TabItem>

<TabItem value="data-result-nodes1">

> Besides graph result in Memgraph Lab, you get the data result - a list of nodes and a list of relationships. Here is the obtained list of nodes:

Reviewer: ```suggestion
Memgraph Lab can also return data results - a list of nodes and a list of relationships. Here is the obtained list of nodes:
```

  ]
}>
  <TabItem value="graph">

  Create a graph by running the following Cypher query:

```cypher
CREATE (n:Person {name: "Kate", age: 27})-[:IS_FRIENDS_WITH]->(m:Person:Student {name: "James", age: 30, year: "second"})-[:STUDIES_AT]->(:University {name: "University of Vienna"})
WITH n, m
CREATE (n)-[:LIVES_IN]->(:City {name: "Zagreb"})<-[:LIVES_IN]-(m);
```
  </TabItem>

  <TabItem value="run">

Once the graph is created, run the following code to call the `schema` procedure:

```cypher
CALL meta_util.schema(true) 
YIELD nodes, relationships 
RETURN nodes, relationships;
```

  </TabItem>

  <TabItem value="graph-result">

The graph result of the `schema` procedure can be seen in Memgraph Lab, and it looks like this:
<!-- 
<div style={{textAlign:"center"}}> -->
<div className={"imgRow"}>
<div className={"imgColumn"}>
<img src={require('../../data/query-modules/python/meta_util/meta_result.png').default}/>
</div>
<div className={"imgColumn"}>
<img src={require('../../data/query-modules/python/meta_util/meta_result_count.png').default}/>
</div>
</div>
  </TabItem>

<TabItem value="data-result-nodes">

> Besides graph result in Memgraph Lab, you get the data result - a list of nodes and a list of relationships. Here is the obtained list of nodes:

Reviewer: also change like in the comment above

---
title: meta_util
sidebar_label: meta_util
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
import RunOnSubgraph from '../../templates/_run_on_subgraph.mdx';

export const Highlight = ({children, color}) => (
  <span
    style={{
      backgroundColor: color,
      borderRadius: '2px',
      color: '#fff',
      padding: '0.2rem',
    }}>
    {children}
  </span>
);

<!-- TODO -- update link -->
[![docs-source](https://img.shields.io/badge/source-meta_util-FB6E00?logo=github&style=for-the-badge)](https://github.com/memgraph/mage/blob/main/python/meta_util.py)

## Abstract

A module that contains procedures describing graphs on a meta-level.

| Trait               | Value                                                 |
| ------------------- | ----------------------------------------------------- |
| **Module type**     | <Highlight color="#FB6E00">**util**</Highlight>  |
| **Implementation**  | <Highlight color="#FB6E00">**Python**</Highlight>     |
| **Parallelism**     | <Highlight color="#FB6E00">**sequential**</Highlight> |

## Procedures

<RunOnSubgraph/>

### `schema(include_properties)`

> Knowing what kind of data, that is, what kind of nodes and relationships, is stored inside the database and how they're connected can be helpful. Besides that, each node or relationship can have a set of properties, and while loading the data in the database, you should be sure that a certain amount of graph objects has a particular property. That's where the number of graph objects with a particular property (property count) might come in handy. 

Reviewer: ```suggestion
Knowing what kind of data, that is, what kind of nodes and relationships, are stored inside the database and how they're connected can be helpful. Besides that, each node or relationship can have a set of properties, and while loading the data in the database, you should be sure that a certain amount of graph objects has a particular property. That's where the number of graph objects with a particular property (property count) might come in handy. 
```
